hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateNftsHandler.java

```diff
@@ -0,0 +1,146 @@
+
+import com.hedera.hapi.node.base.SubType;
+import com.hedera.hapi.node.base.TokenID;
+import com.hedera.hapi.node.state.token.Nft;
+import com.hedera.hapi.node.token.TokenUpdateNftsTransactionBody;
+import com.hedera.hapi.node.transaction.TransactionBody;
+import com.hedera.node.app.service.token.ReadableTokenStore;
+import com.hedera.node.app.service.token.impl.WritableNftStore;
+import com.hedera.node.app.service.token.impl.validators.TokenAttributesValidator;
+import com.hedera.node.app.spi.fees.FeeContext;
+import com.hedera.node.app.spi.fees.Fees;
+import com.hedera.node.app.spi.workflows.HandleContext;
+import com.hedera.node.app.spi.workflows.HandleException;
+import com.hedera.node.app.spi.workflows.PreCheckException;
+import com.hedera.node.app.spi.workflows.PreHandleContext;
+import com.hedera.node.app.spi.workflows.TransactionHandler;
+import com.hedera.node.config.data.TokensConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * Provides the state transition for an NFT collection update.
+ */
+@Singleton
+public class TokenUpdateNftsHandler implements TransactionHandler {
+    private final TokenAttributesValidator validator;
+
+    @Inject
+    public TokenUpdateNftsHandler(@NonNull final TokenAttributesValidator validator) {
+        this.validator = validator;
+    }
+
+    @Override
+    public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckException {
+        requireNonNull(txn);
+        final var op = txn.tokenUpdateNftsOrThrow();
+        validateTruePreCheck(op.hasToken(), INVALID_TOKEN_ID);
+        validateTrue(!op.serialNumbers().isEmpty(), MISSING_SERIAL_NUMBERS);
+    }
+
+    @Override
+    public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {
+        requireNonNull(context);
+        final var txn = context.body();
+        final var op = txn.tokenUpdateNftsOrThrow();
+        final var tokenStore = context.createStore(ReadableTokenStore.class);
+        final var token = tokenStore.get(op.tokenOrElse(TokenID.DEFAULT));
+        if (token == null) throw new PreCheckException(INVALID_TOKEN_ID);
+        if (token.hasMetadataKey()) {
+            context.requireKey(token.metadataKeyOrThrow());
+        }
+    }
+
+    @Override
+    public void handle(@NonNull HandleContext context) throws HandleException {
+        requireNonNull(context);
+        final var txnBody = context.body();
+        final var op = txnBody.tokenUpdateNftsOrThrow();
+        final var tokenId = op.tokenOrThrow();
+
+        // Ensure that the token has metadataKey
+        final var tokenStore = context.readableStore(ReadableTokenStore.class);
+        final var token = getIfUsable(tokenId, tokenStore);
+        validateTrue(token.hasMetadataKey(), TOKEN_HAS_NO_METADATA_KEY);
+
+        if (!op.hasMetadata()) {
+            return;
+        }
```

Comment:
q: If a user sets metadata to empty and serials to invalid values (-1, max long, etc) or non-existing serials, is it okay to not validate the serials and return success for no-op transaction?

Reply:
The same we do in similar cases elsewhere. I can check once I am back at my desk.

Reply:
I think we validate in other handlers. But we were thinking if we should lookup state. I will update to fail if the nft is not valid.